app/models/manageiq/providers/amazon/cloud_manager/ssa_agent_manager.rb

```diff
@@ -0,0 +1,326 @@
+    "echo -e '---\n:log_level: #{log_level}\n:region: #{region}\n:request_queue: #{request_queue}\n:reply_queue: #{reply_queue}\n:ssa_bucket: #{ssa_bucket}\n:reply_prefix: #{reply_prefix}\n:log_prefix: #{log_prefix}\n:heartbeat_prefix: #{heartbeat_prefix}\n:heartbeat_interval: i#{heartbeat_interval}\n' > default_ssa_config.yml"
+  end
+
+  def create_user_data
+    script = '#!/bin/bash
+yum -y update > /var/log/miq_ssa_deploy.log 2>&1
+yum -y install git-core zlib zlib-devel gcc-c++ patch readline readline-devel libyaml-devel libffi-devel openssl-devel make bzip2 autoconf automake libtool bison curl sqlite-devel postgresql-devel >> /var/log/miq_ssa_deploy.log 2>&1
+git clone https://github.com/rbenv/rbenv.git ~/.rbenv >> /var/log/miq_ssa_deploy.log 2>&1
+export HOME="/root" >> /var/log/miq_ssa_deploy.log 2>&1
+echo \'export PATH="$HOME/.rbenv/bin:$PATH"\' >> ~/.bash_profile
+echo \'eval "$(/root/.rbenv/bin/rbenv init -)"\' >> ~/.bash_profile
+source ~/.bash_profile >> /var/log/miq_ssa_deploy.log 2>&1
+git clone git://github.com/sstephenson/ruby-build.git ~/.rbenv/plugins/ruby-build >> /var/log/miq_ssa_deploy.log 2>&1
+echo \'export PATH="$HOME/.rbenv/plugins/ruby-build/bin:$PATH"\' >> ~/.bash_profile
+source ~/.bash_profile >> /var/log/miq_ssa_deploy.log 2>&1
+echo $PATH >> /var/log/miq_ssa_deploy.log 2>&1
+rbenv install -l >> /var/log/miq_ssa_deploy.log 2>&1
+rbenv install 2.3.3 >> /var/log/miq_ssa_deploy.log 2>&1
+rbenv global 2.3.3 >> /var/log/miq_ssa_deploy.log 2>&1
+gem install bundler >> /var/log/miq_ssa_deploy.log 2>&1
+gem install rails >> /var/log/miq_ssa_deploy.log 2>&1
+gem install aws-sdk >> /var/log/miq_ssa_deploy.log 2>&1
+rbenv rehash >> /var/log/miq_ssa_deploy.log 2>&1
+ruby -v >> /var/log/miq_ssa_deploy.log 2>&1
+mkdir -p /opt/miq/log
+cd /opt/miq
+echo \'source "https://rubygems.org"
+gem "manageiq-gems-pending", ">0", :require => "manageiq-gems-pending", :git => "https://github.com/ManageIQ/manageiq-gems-pending.git", :branch => "master"
+gem "manageiq-smartstate", ">0", :require => "manageiq-smartstate", :git => "https://github.com/ManageIQ/manageiq-smartstate.git", :branch => "master"
+gem "amazon_ssa_support", ">0", :require => "amazon_ssa_support", :git => "https://github.com/ManageIQ/amazon_ssa_support.git", :branch => "master"
+# Modified gems for gems-pending.  Setting sources here since they are git references
+gem "handsoap", "~>0.2.5", :require => false, :git => "https://github.com/ManageIQ/handsoap.git", :tag => "v0.2.5-5"
+gem "aws-sdk"\' > Gemfile
+bundle install >> /var/log/miq_ssa_deploy.log 2>&1
+' + default_settings + '
+echo \'#!/bin/sh\' > start_agent.sh
+echo \'ssa_root=`bundle show amazon_ssa_support`\' >> start_agent.sh
+echo \'ssa_script="$ssa_root/tools/amazon_ssa_extract.rb"\' >> start_agent.sh
+echo \'ruby ${ssa_script}\' >> start_agent.sh
+chmod 755 start_agent.sh
+echo "Agent starts ...." >> /var/log/miq_ssa_deploy.log 2>&1
+./start_agent.sh
+'
+    Base64.encode64(script)
+  end
+
```

Comment:
This seems like a lot of shell script code to have defined inside a ruby source file.
Two ideas - and the second one is better - 
1) The script should be defined via a "Here document" instead of just a really long string.
2) The script should be an external file defined in the source tree.